buildSrc/src/main/java/org/elasticsearch/gradle/info/GenerateGlobalBuildInfoTask.java

```diff
@@ -0,0 +1,270 @@
+        String versionInfoScript = "print(" +
+            "java.lang.System.getProperty(\"java.vendor\") + \" \" + java.lang.System.getProperty(\"java.version\") + " +
+            "\" [\" + java.lang.System.getProperty(\"java.vm.name\") + \" \" + java.lang.System.getProperty(\"java.vm.version\") + \"]\");";
+        return runJavaAsScript(javaHome, versionInfoScript).trim();
+    }
+
+    /**
+     * Finds the parsable java specification version
+     */
+    private String findJavaSpecificationVersion(File javaHome) {
+        String versionScript = "print(java.lang.System.getProperty(\"java.specification.version\"));";
+        return runJavaAsScript(javaHome, versionScript);
+    }
+
+    /**
+     * Runs the given javascript using jjs from the jdk, and returns the output
+     */
+    private String runJavaAsScript(File javaHome, String script) {
+        ByteArrayOutputStream stdout = new ByteArrayOutputStream();
+        ByteArrayOutputStream stderr = new ByteArrayOutputStream();
+        if (OS.current() == OS.WINDOWS) {
+            // gradle/groovy does not properly escape the double quote for windows
+            script = script.replace("\"", "\\\"");
+        }
+        File jrunscriptPath = new File(javaHome, "bin/jrunscript");
+        String finalScript = script;
+        ExecResult result = getProject().exec(spec -> {
+            spec.setExecutable(jrunscriptPath);
+            spec.args("-e", finalScript);
+            spec.setStandardOutput(stdout);
+            spec.setErrorOutput(stderr);
+            spec.setIgnoreExitValue(true);
+        });
+
+        if (result.getExitValue() != 0) {
+            getLogger().error("STDOUT:");
+            Arrays.stream(stdout.toString(UTF_8).split(System.getProperty("line.separator"))).forEach(getLogger()::error);
+            getLogger().error("STDERR:");
+            Arrays.stream(stderr.toString(UTF_8).split(System.getProperty("line.separator"))).forEach(getLogger()::error);
+            result.rethrowFailure();
+        }
+        return stdout.toString(UTF_8).trim();
+    }
+
+    @Input
+    public JavaVersion getMinimumCompilerVersion() {
+        return minimumCompilerVersion;
+    }
+
+    public void setMinimumCompilerVersion(JavaVersion minimumCompilerVersion) {
+        this.minimumCompilerVersion = minimumCompilerVersion;
+    }
+
+    @Input
+    public JavaVersion getMinimumRuntimeVersion() {
+        return minimumRuntimeVersion;
+    }
+
+    public void setMinimumRuntimeVersion(JavaVersion minimumRuntimeVersion) {
+        this.minimumRuntimeVersion = minimumRuntimeVersion;
+    }
+
+    @InputDirectory
+    @PathSensitive(PathSensitivity.RELATIVE)
+    public File getCompilerJavaHome() {
+        return compilerJavaHome;
+    }
+
+    public void setCompilerJavaHome(File compilerJavaHome) {
+        this.compilerJavaHome = compilerJavaHome;
+    }
+
+    @InputDirectory
+    @PathSensitive(PathSensitivity.RELATIVE)
```

Comment:
I'm not sure this would work with relative for example we have `~/.java/java11` in CI. My understanding is that this would use `java11` as a cache key, but patch level can change  so in one point in time we might have `11.0.1` and then `11.0,.2` at that location. 
I think up to date checks are fine, since something will surely change in the directory for sure, but not if something is pulled from the cache.

Reply:
This path sensitivity means that we don't care about the root directory name. Only the paths _below_ that directory. 

For example. Let's say my java home is `~/.java/java11.0.1` and there exists a file `~/.java/java11.0.1/bin/java`. If that file were to move to `~/.java/java11.0.1/jre/bin/java` that would be considered a change in the cache key. However, if I were to simply rename `~/.java/java11.0.1` to `~/.java/java` no _relative_ paths would change so no change in the cache key.

So in CI if `~/.java/java11` and `~/.java/java` both point to the same place, they will both have identical cache keys since the actual root folder name is not considered.

Also, to clarify, the actual _contents_ of this directory are also hashed so any change there would invalidate the cache entry.